### DISCUSSION TITLE
GPU instancing

### DISCUSSION THREAD

**Discussion by nowakpe** [nxrighthere/UnrealCLR#47] [Help] [answered]
Is there a method similar to Unity "https://docs.unity3d.com/ScriptReference/Graphics.DrawMeshInstanced.html"
I could not find a similar simple method directly in UE4, so I am expecting that it is not, but if someone has any ideas? Thanks.

Update:
I found it.
`UInstancedStaticMeshComponent::BatchUpdateInstancesTransforms`
But it is not yet exposed. 
Right now only for single instance `InstancedStaticMeshComponent.UpdateInstanceTransform`

**Comment by nxrighthere** [comment, marked as answer]
`UInstancedStaticMeshComponent::BatchUpdateInstancesTransforms()` is essentially doing the same thing under the hood. Just use `markRenderStateDirty` for the last element in a batch with `InstancedStaticMeshComponent.UpdateInstanceTransform()`, as stated in the [API reference](https://github.com/nxrighthere/UnrealCLR/blob/master/API/InstancedStaticMeshComponent-UpdateInstanceTransform(int_Transform_bool_bool_bool).md). See [this](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/InstancedStaticMeshes.cs) test.

**Reply by nowakpe** [reply]
Thanks, I completely missed the functionality of that argument. 

But even when it is doing the same thing under the hood, it is just one method call instead of thousands in one frame.
I was trying to add that method by myself, but it is crashing the editor, is it even possible to add something by myself, or do I need also to modify something behind the scenes? (It compiled and was started without any errors) 

One more thing which is a little bit strange, let's say for 40K instances it gives me about 5 fps more if instead of
`UpdateInstanceTransform(...)` in for loop
I do
`ClearInstances()`
and in for loop, I will add them again` AddInstance(...)`

**Reply by nxrighthere** [reply]
> But even when it is doing the same thing under the hood, it is just one method call instead of thousands in one frame.

The same thing under the hood, it's just a helper function that iterating over an array.

> I was trying to add that method by myself, but it is crashing the editor, is it even possible to add something by myself, or do I need also to modify something behind the scenes? (It compiled and was started without any errors)

You can expose native functionality, but you need to know what you are doing. Interop might be very tricky.

> One more thing which is a little bit strange, let's say for 40K instances it gives me about 5 fps more if instead of

Most likely, the update method has additional checks. To measure performance, you have to use a standalone build with a shipping configuration.

**Reply by nowakpe** [reply]
> The same thing under the hood, it's just a helper function that iterating over an array.

I know but still, you do only one native call instead of thousands. I will give one more try to get running the "Batch" method,
I will then let you know the difference.

> Most likely, the update method has additional checks. To measure performance, you have to use a standalone build with a shipping configuration.

True, and if we are at builds... :-D I have another problem. Build runs successfully, but when I start, it throws an error:
"Plugin UnrealCLR failed to load because the module UnrealCLR could not be found."

I could not find s proper solution to this, nothing I found helped, any idea?

**Reply by nxrighthere** [reply]
> I could not find s proper solution to this, nothing I found helped, any idea?

Package a project with a development configuration, run it, and then check log files in `Saved` folder.

**Reply by nxrighthere** [reply]
> Plugin UnrealCLR failed to load because the module UnrealCLR could not be found

Also, see https://answers.unrealengine.com/questions/987779/view.html

**Reply by nxrighthere** [reply]
> I know but still, you do only one native call instead of thousands.

[Exposed](https://github.com/nxrighthere/UnrealCLR/blob/master/API/InstancedStaticMeshComponent-BatchUpdateInstanceTransforms(int_Transform--_bool_bool_bool).md) in the master.

**Reply by nowakpe** [reply]
> Also, see https://answers.unrealengine.com/questions/987779/view.html

Thank you very much, this helped!

> Exposed in the master.

I checked the implementation... now I see what I forgot to get my version working :-D 
This line:
` InstancedStaticMeshComponent.batchUpdateInstanceTransforms = (delegate* unmanaged[Cdecl]<IntPtr, int, int, Transform[], Bool, Bool, Bool, Bool>)instancedStaticMeshComponentFunctions[head++];`

Anyway, I am using my version as it gives me about 20fps more at 120K instances (built):
[batch1](https://imgur.com/a/Nk5BnAk)
[batch2](https://imgur.com/a/K6oodjh)

I had to also create a new struct for Transform to be able to do that as Unreal is using 4 float vectors.

**Reply by nowakpe** [reply]
Maybe also if I wanted to go more deeply I could probably expose directly pointer to the data of that `TArray` to prevent copying at all.
But that is enough for now, I was just playing around. :-) And I am quite surprised that even without instancing it could handle like 7K Actors at 60 fps. (Of course, it depends also on HW )

**Reply by nxrighthere** [reply]
> Anyway, I am using my version as it gives me about 20fps more at 120K instances (built):

Is that in comparison to the original version without memory copying?

**Reply by nxrighthere** [reply]
> I had to also create a new struct for Transform to be able to do that as Unreal is using 4 float vectors.

Are you sure that your code is executing correctly? `FTransform` has at least 3 fields with rotation, translation, and scale. It's also 16-byte aligned, so 48 bytes in size. How a transform with 4 float fields matching to that layout?

**Reply by nowakpe** [reply]
> Is that in comparison to the original version without memory copying?

Yes, but if you used directly c++ in Unreal you do not have to copy anything also, it is just because that method accepts the type of TArray.

> Are you sure that your code is executing correctly? FTransform has at least 3 fields with rotation, translation, and scale. It's also 16-byte aligned, so 48 bytes in size. How a transform with 4 float fields matching to that layout?

Sorry for the misunderstanding, to be clearer. By "4 float vectors" I meant that the FTransform is Quaternion (4 floats), Location (4 floats), and Scale (4 floats).
So I am using this:

`	struct Transform4 { Quaternion Rotation; Vector4 Location; Vector4 Scale;}`

Maybe it can be also achieved with Vector3 and explicit layout? It is probably more performant to use 16 bytes, because of the vectorization, SIMD stuff...

**Reply by nxrighthere** [reply]
> Maybe it can be also achieved with Vector3 and explicit layout? It is probably more performant to use 16 bytes, because of the vectorization, SIMD stuff...

I have no control over .NET vector types from `System.Numerics`, so intermediate vector types are used to map them into Unreal vector types.

> By "4 float vectors" I meant that the FTransform is Quaternion (4 floats), Location (4 floats), and Scale (4 floats).

Ah, that makes sense yes.

> Yes, but if you used directly c++ in Unreal you do not have to copy anything also, it is just because that method accepts the type of TArray.

Try to use this constructor:

```cpp
TArray(const ElementType* Ptr, SizeType Count)
```

**Reply by nowakpe** [reply]
> Try to use this constructor:

Checked, it is also using memcpy internally. But thanks for the tip.

Edit: Hm ok, I also tried it and for 150K instances, it gives another 10 fps :D maybe I missed some branching.

**Reply by nowakpe** [reply]
The result https://www.linkedin.com/posts/petr-nowak-513198156_unrealengine-ue4-unrealclr-activity-6723503591308464128-kZyp
(Didn't know that linked in will decrease the quality of the video that much :-D)

**Reply by nxrighthere** [reply]
Pretty cool! May you share it here in the `Showcase` category, please?

Yes, all those websites are screwing video quality.

**Reply by nxrighthere** [reply]
Transforms are now used directly where appropriate https://github.com/nxrighthere/UnrealCLR/commit/8aa5c6e162a89a94d3db933313acebea1cdd31cc. Thanks for participation.
